components/avatar/avatar.vue

```diff
@@ -0,0 +1,108 @@
+<template>
+  <span :class="classString">
+    <img v-if="src && isImgExist" :src="src" @error="handleImgLoadError" />
+    <Icon v-else-if="icon" :type="icon" />
+    <span v-else ref="children" :class="`${ prefixCls }-string`" :style="childrenStyle">
+      <slot></slot>
+    </span>
+  </span>
+</template>
+
+<script>
+import Icon from '../icon'
+import { oneOf } from '../_util/proptype'
+
+const SIZEMAP = {
+  'small': 'sm',
+  'large': 'lg'
+}
+
+export default {
+  name: 'avatar',
+  props: {
+    icon: String,
+    prefixCls: {
+      type: String,
+      default: 'ant-avatar'
+    },
+    shape: {
+      validator: function (value) {
+        return oneOf(value, ['circle', 'square'])
+      },
+      default: 'circle'
+    },
+    size: {
+      validator: function (value) {
+        return oneOf(value, ['large', 'small', 'default'])
+      },
+      default: 'default'
+    },
+    src: {
+      type: String
+    }
+  },
+  data () {
+    return {
+      scale: 1,
+      isImgExist: true,
+      offsetWidth: 0
+    }
+  },
+  computed: {
+    childrenStyle () {
+      return {
+        transform: `scale(${this.scale})`,
+        position: 'absolute',
+        display: 'inline-block',
+        left: `calc(50% - ${Math.round(this.offsetWidth / 2)}px)`
+      }
+    },
+    classString () {
+      const prefixCls = this.prefixCls
+      const size = this.size && SIZEMAP[this.size]
+      return [
+        `${prefixCls}`,
+        {
+          [`${prefixCls}-${size}`]: !!size,
+          [`${prefixCls}-${this.shape}`]: !!this.shape,
+          [`${prefixCls}-image`]: !!this.src,
+          [`${prefixCls}-icon`]: !!this.icon
+        }
+      ]
+    }
+  },
+  components: {
+    Icon
+  },
+  methods: {
+    setScale () {
+      const childrenNode = this.$refs.children
+      if (childrenNode) {
+        // scale can`t affect offsetWidth, but affect actual size（or computed size）
+        const childrenWidth = this.offsetWidth
+        const avatarWidth = this.$el.getBoundingClientRect().width
+
+        if (avatarWidth - 8 < childrenWidth) {
+          this.scale = (avatarWidth - 8) / childrenWidth
+        } else {
+          this.scale = 1
+        }
+      }
+    },
+    handleImgLoadError () {
+      this.isImgExist = false
+    },
+    getOffSet () {
```

Comment:
It seems no need to make getOffSet to be a function, can be in setScale, and rename it 'initAndUpdateAvatar'. just my opinion, not a problem.

Reply:
getOffSet的目的是更新依赖于offsetWidth的childrenStyle。因为原来使用 ‘this.$refs.children.offsetWidth’ 的时候，childrenStyle不会更新，所以才增加在data中增加了offsetWidth并且增加相应的函数进行更新它。不过我感觉我应该把函数名字改成 ‘updateOffSetWidth’。

Reply:
The purpose of getOffSet is to update childrenStyle that relies on offsetWidth. Because the original use of 'this.$refs.children.offsetWidth', childrenStyle will not be updated, so it is added to increase the offsetWidth in data and add the corresponding function to update it. But I feel I should change the name of the function to 'updateOffSetWidth'.